src/ngx_http_lua_kong_request.c

```diff
@@ -0,0 +1,109 @@
+#include <ngx_config.h>
+#include <ngx_core.h>
+#include <ngx_http.h>
+
+/* name should be lowercased and preprocessed for both search functions */
+
+/* by hash */
+static ngx_str_t *search_known_header(ngx_http_request_t *r, ngx_str_t name)
+{
+    /* Calculate a hash of lowercased header name */
+    ngx_uint_t hash = 0;
+    for (size_t i = 0; i < name.len; i++) {
+        hash = ngx_hash(hash, name.data[i]);
+    }
+
+    ngx_http_core_main_conf_t *cmcf =
+        ngx_http_get_module_main_conf(r, ngx_http_core_module);
+    ngx_http_header_t *hh =
+        ngx_hash_find(&cmcf->headers_in_hash, hash, name.data, name.len);
+
+    /* There header is unknown or is not hashed yet. */
+    if (hh == NULL) {
+        return NULL;
+    }
+
+    /* There header is hashed but not cached yet for some reason. */
+    if (hh->offset == 0) {
+        return NULL;
+    }
+
+    /* The header value was already cached in some field of the r->headers_in
+        struct (hh->offset tells in which one). */
+    return &((*(ngx_table_elt_t **)
+        ((char *) &r->headers_in + hh->offset))->value);
+}
+
+/* linear search */
+static ngx_str_t *search_unknown_header(
```

Comment:
The function's name should have a prefix` ngx_http_`.

Reply:
Ditto